composer.json

```diff
@@ -49,6 +49,9 @@
 		"phpunit/phpunit": "^10.1",
 		"rector/rector": "^0.15.0 || ^0.16.0 || ^0.17.0"
 	},
+	"suggest": {
+		"gmostafa/php-graphql-client": "Used to perform GraphQL requests within GithubVcsProvider (^1.13)"
+	},
```

Comment:
I suppose it could be inconvenient when trying to instantiate https://github.com/CPS-IT/frontend-asset-handler/blob/d77066d348864ec35ae9cc2b8d446f00ed0bf229/src/Vcs/GithubVcsProvider.php#L30 when the requirement is only suggested, especially if users don't know about the implication of not following the composer suggestion.

Reply:
You're right. We should probably add a `class_exists()` check in front and throw an explanatory exception suggesting the additional `composer require`. What do you think?

Reply:
On it.

Reply:
Not so fast there. I'm afraid we need to discuss this further. I'm not really happy with the solution I came up with.

My initial urge was to skip out early in the Factory:

```php
public function get(string $type, Asset\Definition\Vcs $vcs = null): VcsProviderInterface
// ...
    if (!$this->typeSuggestedClassExists($type)) {
       throw Exception\MissingClassException::create($type);
    }
// ...
}

private function typeSuggestedClassExists(string $type): bool
{
    return match ($type) {
        'github' => class_exists(\GraphQL\Client::class),
        default => true,
    };
}

```

Then again, I obviously would love to give some context in the exception so I would have to print the class-strings there, too. But I cannot seem to find a suitable solution that doesn't feel hacky and dirty.

Why would we create such a detour just for a class that requires something that we're only suggesting. Should we perhaps skip out in the instantiation? That would detach the whole "special treatment" from the Factory and keep it in the affected Provider.

```php
final class GithubVcsProvider implements DeployableVcsProviderInterface
{
    // ...
    public function __construct(
        private readonly ClientInterface $client,
        private ?string $accessToken = null,
        private ?string $owner = null,
        private ?string $name = null,
        private ?string $environment = null,
    ) {

        if (
            !class_exists(\GraphQL\Client::class) ||
            // ...
        ) {
			throw Exception\MissingClassException::create($type, $className);
		}
    }
````

Ya, see? We need to talk ;-)